xarray/tests/test_datatree.py

```diff
@@ -666,20 +690,229 @@ def test_repr(self):
+        xr.testing.assert_equal(dt["/b/z"], xr.DataArray(3, coords=sub_coords))
+
+    def test_inconsistent_dims(self):
+        expected_msg = (
+            "^"
+            + re.escape(
+                dedent(
+                    """
+                group '/b' is not aligned with its parent:
+                Group:
+                    Dimensions:  (x: 1)
+                    Dimensions without coordinates: x
+                    Data variables:
+                        c        (x) int64 8B 3
+                Parent:
+                    Dimensions:  (x: 2)
+                    Dimensions without coordinates: x
+                """
+                ).strip()
+            )
+            + "$"
+        )
+
+        with pytest.raises(ValueError, match=expected_msg):
+            DataTree.from_dict(
+                {
+                    "/": xr.Dataset({"a": (("x",), [1, 2])}),
+                    "/b": xr.Dataset({"c": (("x",), [3])}),
+                }
+            )
+
+        dt: DataTree = DataTree()
+        dt["/a"] = xr.DataArray([1, 2], dims=["x"])
+        with pytest.raises(ValueError, match=expected_msg):
+            dt["/b/c"] = xr.DataArray([3], dims=["x"])
+
+        b: DataTree = DataTree(data=xr.Dataset({"c": (("x",), [3])}))
+        with pytest.raises(ValueError, match=expected_msg):
+            DataTree(
+                data=xr.Dataset({"a": (("x",), [1, 2])}),
+                children={"b": b},
+            )
+
+    def test_inconsistent_child_indexes(self):
+        expected_msg = (
+            "^"
+            + re.escape(
+                dedent(
+                    """
+                group '/b' is not aligned with its parent:
+                Group:
+                    Dimensions:  (x: 1)
+                    Coordinates:
+                      * x        (x) int64 8B 2
+                    Data variables:
+                        *empty*
+                Parent:
+                    Dimensions:  (x: 1)
+                    Coordinates:
+                      * x        (x) int64 8B 1
+                """
+                ).strip()
+            )
+            + "$"
+        )
+
+        with pytest.raises(ValueError, match=expected_msg):
+            DataTree.from_dict(
+                {
+                    "/": xr.Dataset(coords={"x": [1]}),
+                    "/b": xr.Dataset(coords={"x": [2]}),
+                }
+            )
+
+        dt: DataTree = DataTree()
+        dt.ds = xr.Dataset(coords={"x": [1]})  # type: ignore
+        dt["/b"] = DataTree()
+        with pytest.raises(ValueError, match=expected_msg):
+            dt["/b"].ds = xr.Dataset(coords={"x": [2]})
+
+        b: DataTree = DataTree(xr.Dataset(coords={"x": [2]}))
+        with pytest.raises(ValueError, match=expected_msg):
+            DataTree(data=xr.Dataset(coords={"x": [1]}), children={"b": b})
+
+    def test_inconsistent_grandchild_indexes(self):
```

Comment:
These two `grandchild` tests don't work because of some as-yet unsolved bug in my implementation.

Reply:
Probably this https://github.com/pydata/xarray/pull/9063#issuecomment-2197281807